### ISSUE TITLE
BufferSerializer leaks old data through padding bytes

### ISSUE THREAD

**Issue by Diomendius** [rkyv/rkyv#227]
With default configuration, `Box::<str>::from("0")` serializes to `{ data: b"0", pad: [?_u8; 3], ptr: -4_i32, len: 1_u32 }`. If a `WriteSerializer` is used, the 3 padding bytes are written as `\0`, but if a `BufferSerializer` is used, the padding bytes are not written to at all, leaving their original values in the buffer. This is surprising, especially as the serialized output depends on the type of serializer used.

This pad-by-seek feature of `BufferSerializer` could be desirable in some cases, but I would expect pad-with-zeroes to be both the desired and expected behaviour in most cases. It is not currently feasible to avoid this issue without explicitly overwriting the buffer each time it is reused.

The following code illustrates the issue:
```rust
use std::convert::TryInto;

use rkyv::ser::{
    serializers::{BufferSerializer, WriteSerializer},
    Serializer,
};

fn print_buf(p: &[u8]) {
    println!(
        "data: {:?}, ptr: {}, len: {}",
        &p[0..4],
        i32::from_ne_bytes(p[4..8].try_into().unwrap()),
        u32::from_ne_bytes(p[8..12].try_into().unwrap()),
    );
}

fn main() {
    let mut buf = [10_u8; 12];
    let s = Box::<str>::from("0");

    let mut ser = BufferSerializer::new(&mut buf);
    ser.serialize_value(&s).unwrap();
    print_buf(&buf);

    buf.fill(10);

    let mut ser = WriteSerializer::new(&mut buf[..]);
    ser.serialize_value(&s).unwrap();
    print_buf(&buf);
}
```

The cause is here, as `BufferSerializer` explicitly seeks itself in its implementation of `pad()`: https://github.com/rkyv/rkyv/blob/0a9394b61d03a049e85b58b1b2b8de70835ed812/rkyv/src/ser/serializers/core.rs#L161

**Comment by djkoloski** [comment]
While I agree that this behavior may be unintuitive, it seems okay to me. If the buffer is already zeroed then there's no point in re-zeroing the memory, and not doing so would save cycles. `WriteSerializer` explicitly writes zeroes for padding because it needs to adapt to the underlying writer.

I'm a little on the fence with this one, but in the end I think that your suggestion would be a good idea. If I switched from a `WriteSerializer` to a `BufferSerializer`, it would violate the principle of least surprise if the posting bytes came out differently. So overall, I'm in favor of this change.

**Comment by Diomendius** [comment]
Yes, this is not an issue if writing to a newly-created zero-initialized buffer. If the buffer is reused, bytes from previous serializations leak through, which isn't a serious issue in the way leaking uninitialized data is, but would cause problems if, for instance, the serialized data were hashed as a unit. Another contrived but plausible scenario is if the same buffer were reused to send data to different destinations that shouldn't have read access to each other.

Ideally, users should have control over whether padding bytes are overwritten, but I have no idea what such an interface should look like.

There's one case where not overwriting the padding bytes can actually cause worse code generation, though I doubt it applies to the real world. When testing, I serialized a struct containing a `&str` and some integer types, all from compile time constant values. The string's length wasn't 4-byte aligned, and the compiler generated an unaligned 64-bit store shifted back to overlap previously written bytes and avoid overwriting the padding. It's actually a pretty clever optimization to avoid multiple smaller stores, but in this case if the compiler were allowed to overwrite the padding, the entire serialization operation would have become an inlined `memcpy`.

I don't think this applies to serialization of dynamic data, and the `copy` and `copy_unsafe` features are more applicable in this context anyway.

**Comment by djkoloski** [comment]
Thanks for the example, this kind of sub-optimal codegen is a known issue with `WriteSerializer` and `BufferSerializer` since everything boils down to byte copies. `AlignedSerializer` uses exclusively `resolve_aligned` when possible which helps codegen a bit.

**Comment by djkoloski** [comment]
Fixed by 3077e3cc7476c742a926a5418d9e5c08564de2b2
